compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -987,14 +987,49 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                 GenericArgs::AngleBracketed(data) => {
                     self.lower_angle_bracketed_parameter_data(data, ParamMode::Explicit, itctx).0
                 }
+                &GenericArgs::ReturnTypeNotation(span) => GenericArgsCtor {
+                    args: Default::default(),
+                    bindings: &[],
+                    parenthesized: hir::GenericArgsParentheses::ReturnTypeNotation,
+                    span,
+                },
                 GenericArgs::Parenthesized(data) => {
-                    self.emit_bad_parenthesized_trait_in_assoc_ty(data);
-                    self.lower_angle_bracketed_parameter_data(
-                        &data.as_angle_bracketed_args(),
-                        ParamMode::Explicit,
-                        itctx,
-                    )
-                    .0
+                    if let Some(start_char) = constraint.ident.as_str().chars().next()
+                        && start_char.is_ascii_lowercase()
+                        && self.tcx.sess.is_nightly_build()
```

Comment:
A we able to check if the feature is enabled instead, or is this happening before we've looked at features?

Reply:
Is there harm in always allowing this to parse, as long as we later gate it with features? It'd nice not to have to remember to remove `is_nightly_build()` when we go to stabilize this.

Reply:
Sorry to keep commenting on this one line, but it looks like this section is trying to tell two similar syntaxes apart. What's the other one?

Reply:
> Sorry to keep commenting on this one line, but it looks like this section is trying to tell two similar syntaxes apart. What's the other one?

The "other" parenthesized generics, like `Fn(i32) -> i32`.

> Is there harm in always allowing this to parse, as long as we later gate it with features?

Well, we're on an error path. I don't see much use in attempting to lower this as RTN or to suggest `#![feature(return_type_notation)]` if we're not on nightly, though.

> A we able to check if the feature is enabled instead, or is this happening before we've looked at features?

Yes, but this specifically avoids doing that, so if I have code that 1. doesn't have RTN enabled, and 2. uses `T: Trait<method(): Send>` *instead* of `T: Trait<method(..): Send>`, it'll suggest fixing it and enabling the feature gate.